src/menu/stateful-container.js

```diff
@@ -149,11 +154,70 @@ export default class MenuStatefulContainer extends React.Component<
         }
         this.handleEnterKey(event);
         break;
+      default:
+        if (this.props.typeAhead) this.findItemToFocus(event);
+        break;
     }
   };
+  findItemToFocus = (event: KeyboardEvent) => {
+    const rootRef = this.props.rootRef ? this.props.rootRef : this.rootRef;
+    const prevIndex = this.state.highlightedIndex;
+
+    this.charsSoFar += event.key;
+    this.clearCharsSoFarAfterDelay();
 
+    var nextIndex = this.findMatchInRange(
+      this.getItems(),
+      prevIndex + 1,
+      this.getItems().length,
+    );
+    if (nextIndex < 0)
+      nextIndex = this.findMatchInRange(this.getItems(), 0, prevIndex);
+    if (nextIndex < 0) nextIndex = prevIndex;
+    this.internalSetState(STATE_CHANGE_TYPES.character, {
+      highlightedIndex: nextIndex,
+    });
+    if (this.refList[nextIndex]) {
+      scrollItemIntoView(
+        this.refList[nextIndex].current,
+        // $FlowFixMe
+        rootRef.current,
+        nextIndex === 0,
+        nextIndex === this.getItems().length - 1,
+      );
+    }
+  };
+  clearCharsSoFarAfterDelay = () => {
+    if (this.charClear) {
+      clearTimeout(this.charClear);
+      this.charClear = null;
+    }
+    this.charClear = setTimeout(
+      function() {
+        this.charsSoFar = '';
+        this.charClear = null;
+      }.bind(this),
+      500,
+    );
+  };
+
+  findMatchInRange = (list: *, startIndex: number, endIndex: number) => {
```

Comment:
I originally did this in your way. But I switch to the two range search method after seeing https://w3c.github.io/aria-practices/examples/listbox/js/listbox.js.
I compare these two methods and found  searching in two separate range is better. My reasons are as follow:
1. people may not interested in content before current focus point. They may be more interested in the contents below current focus point because they haven't seen them. Searching in in two separate range can help search the contents after current point first.
2. If a person is looking for all content with the substring `a` . They can just use keyboard to walk over all the options and press `enter` to select one. If we choose the one you mention, we will all stay in the same place. We cannot allow users to press very long string to perfectly match because we must clean out the string  after some time.

Reply:
I would prefer to mirror default select behavior as shown in this example https://codesandbox.io/s/confident-lumiere-2m64e?file=/index.html

Chrome, safari, and edge all have deterministic selection and firefox works closer to how this is currently implemented

Reply:
Yeah, it's reasonable. I will change it.

Reply:
Yea, I would use the default behavior of Chrome at https://developer.mozilla.org/en-US/docs/Web/HTML/Element/select. It's strange to type item and get the focus moving. Also, they have some sort of fulltext search. (If my first character doesn't match first letter in any item, it's doing fulltext.)

Reply:
yeah, that's a great idea. Will also add fulltext match. Thank you!